db/system_keyspace.cc

```diff
@@ -1653,20 +1650,116 @@ future<std::unordered_map<gms::inet_address, gms::inet_address>> system_keyspace
+    if (opt) {
+        auto set_type = set_type_impl::get_instance(utf8_type, true);
+        opt_values.emplace_back(make_set_value(set_type, prepare_tokens(*opt)));
+        return true;
+    } else {
+        opt_values.emplace_back(std::nullopt);
+        return false;
+    }
+};
+}
+
+future<> system_keyspace::update_peer_info(gms::inet_address ep, peer_info info) {
     if (_db.get_token_metadata().get_topology().is_me(ep)) {
-        on_internal_error(slogger, format("update_peer_info called for this node: {}", ep));
+        //on_internal_error(slogger, format("update_peer_info called for this node: {}", ep));
+        co_return;
+    }
+
+    std::vector<std::optional<data_value>> opt_values;
+    opt_values.reserve(10);
+    bool any_set = false;
+
+    any_set |= ingest_opt_value(opt_values, info.data_center);
+    any_set |= ingest_opt_value(opt_values, info.host_id);
+    any_set |= ingest_opt_value(opt_values, info.preferred_ip);
+    any_set |= ingest_opt_value(opt_values, info.rack);
+    any_set |= ingest_opt_value(opt_values, info.release_version);
+    any_set |= ingest_opt_value(opt_values, info.rpc_address);
+    any_set |= ingest_opt_value(opt_values, info.schema_version);
+    any_set |= ingest_opt_value(opt_values, info.tokens);
+    any_set |= ingest_opt_value(opt_values, info.supported_features);
+
+    if (!any_set) {
+        co_return;
     }
 
-    sstring req = format("INSERT INTO system.{} (peer, {}) VALUES (?, ?)", PEERS, column_name);
-    slogger.debug("INSERT INTO system.{} (peer, {}) VALUES ({}, {})", PEERS, column_name, ep, value);
-    co_await execute_cql(req, ep.addr(), value).discard_result();
+    opt_values.emplace_back(ep);
+
+    auto query = fmt::format("UPDATE system.{} "
+            "SET data_center=?, host_id=?, preferred_ip=?, rack=?, release_version=?, rpc_address=?, schema_version=?, tokens=?, supported_features=? "
+            "WHERE peer=?", PEERS);
+
+    slogger.debug("{}: values={}", query, opt_values);
+
+    co_await _qp.execute_internal(query, db::consistency_level::ONE, opt_values, cql3::query_processor::cache_internal::yes);
+}
+
+// FIXME: implement update_peer_info specializations temporarily,
+// until all callers are converted to use peer_info
+template<> future<> system_keyspace::update_peer_info<sstring>(gms::inet_address ep, sstring column_name, sstring value) {
+    peer_info info;
+    if (column_name == "data_center") {
+        info.data_center.emplace(value);
+    } else if (column_name == "rack") {
+        info.rack.emplace(value);
+    } else if (column_name == "release_version") {
+        info.release_version.emplace(value);
+    } else if (column_name == "supported_features") {
+        info.supported_features.emplace(value);
+    } else {
+        on_fatal_internal_error(slogger, fmt::format("unsupported sstring peer info column_name \"{}\"", column_name));
+    }
+    co_await update_peer_info(ep, info);
```

Comment:
This is really odd, one update_peer_info calling another, but it's not recursion. This is just trying to confuse readers.

Reply:
As I wrote in the comment above, it's just a temporary crutches to reduce code churn and introduce the changes one layer at a time to aid reviewability and testibility of the series.

Reply:
Well it would have been better to use a different name.

Reply:
It's a pre-existing overload (see https://github.com/scylladb/scylladb/blob/f1dea4bc8ad8b443733454aa5abb319c4d1da230/db/system_keyspace.cc#L1661-L1675)
and it's going away in this series, so I don't think it's worth the trouble.

Reply:
ok